File: Data Structures & Algorithms/two-integer-sum/submission-2.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        numMap = {}

        for i in range(0, len(nums)):
            difference = target - nums[i]

            if difference not in numMap:
                numMap[nums[i]] = i
                print(numMap)
            else:
                # print("found difference")
                return [numMap[difference],i]

        print(numMap)
        return []

File: Data Structures & Algorithms/binary-search/submission-2.cs
/*
Input:
    - int[] : nums
    - int : target
Output:
    - int : index of the target, otherwise -1 if not found 

Constraints:
    - O(logn) time 
    - sorted in ascending order 

Plan: 
1. Linear scan - O(n)
2. Binary Search - O(log n)

*/

public class Solution {
    public int Search(int[] nums, int target) {
        // two pointers for search range 
        int left = 0;
        int right = nums.Length - 1;

        // perform binary search
        while (left <= right) {
            // find the middle element 
            int mid = (int)Math.Floor((double)(left + right) / 2);

            // check if the middle element = target
            if (target == nums[mid]) {
                return mid;
            }

            // search left if target < middle element
            if (target < nums[mid]) {
                right = mid - 1;
            }
            // search right if target > middle element
            else {
                left = mid + 1;
            }
        }

        return -1;
    }
}
